docs/Configuration.md
# Configuration

JAR configuration can occur via file or ENV. This documentation assumes *YAML* specifically. JAR can read configuration in JSON, TOML, YAML, HCL, or Java properties -formats. The config variable *names* are case insensitive, but be careful with the *values* as most are strict.

This documentation reflects configuration possibilities as of the associated commit. Please extend this documentation, or ask questions, as they come up.

```bash
Usage of ./jar:
      --checkconfig     Run through the config load and then exit
      --config string   Config file to load
      --debug           Enable vociferous output
      --dumpconfig      Load the config, dump it to stderr, and then exit
      --gopsagent       Start the 'gops' agent
      --version         Print the version and then exit
```

## General

### authoritativedomains: [list]

A list of domains or hostname suffixes this instance will handle. Anything not matching will return 400. This also impacts **tls.httpredirects**.

```yaml
authoritativedomains:
  - .example.com
  - .example.net
  - .example.eu
  - specifically.ahost.com
```

### compression: [list]

A list of MIME types that are eligible for wire-time compression if the client requests it

```yaml
compression:
  - text/xml
  - text/css
  - text/javascript
  - application/javascript
  - application/json
```

### config: [configfile]

Only available via CLI or ENV, points to the configuration file to load. Configuration in JSON, TOML, YAML, HCL, or Java properties -formats is allowed.

### debug: [true|false]

**Default: false**
Also available via CLI as **--debug** this will enable debug logging (see **debuglog** for more) and possibly open additional diagnostic codepaths including spawning goroutines for monitoring and emitting

### debugrequests: [true|false]

**Default: false**
Only enabled if **debug: true** is set.
Enables debug logging of full requests (POST/PUT bodies omitted)

### debugresponses: [true|false]

**Default: false**
Only enabled if **debug: true** is set.
Enables debug logging of full responses (response bodies omitted, generally)

### debugtimings: [true|false] *(Caution)*

**Default: false**
Only enabled if **debug: true** is set.
Enables debug logging of various function timings. Very noisy. **May be removed at any time**.

### ec2: [true|false]

**Default: false**
Enables AWS EC2 awareness.

### fakexfflog: [true|false]

**Default: false**
Backfills the X-Forwarded-For field in the access log, if it isn't already occupied.

### forbiddenpaths: [list]

List of regular expressions that if matched against a URI path, will result in a 403. Barewords will match anywhere in the path. Remember to escape your fore-slashes. *Case-insensitive*.  If you want to limit these to only certain **Paths**, see **Path.ForbiddenPaths**

```yaml
forbiddenpaths:
  - ^\/supersecret\/
  - ^\/(?!admin\/).*healthcheck
  - omgnotthis
```

### handlers: [list]

Handlers listed here will be prepended to the **handlers** list of *all* Paths.

```yaml
handlers:
  - Recoverer
  - ErrorHandler
```

### headers: [list]

Headers listed here will be added to all outgoing responses, unless there is only a name and no value, in which case the header will be *removed* from all responses. Header names are assumed to be everything *left* of the first space. Values are assumed to be everything *right* of the first space. The **headers** list fully allows for **Macros** in the *values*.

```yaml
headers:
  # Set HSTS for TLS
  - "Strict-Transport-Security max-age=31536000"
  # Set the X-JAR header to a macro
  - "X-JAR %%FULLVERSION"
  # Strip the "Server" header from all responses
  - "Server"
```

### keepalivetimeout: [duration]

**Default: 5s**
Specifies the amount of time to allow a keptalive socket to linger.

### listen: [address]

**Default: :8080**
Specifies the *address:port* or *:port* to listen for HTTP requests on.

**NOTE:** Changing this value while running will create problems for graceful restarts, and should only be done during controlled stop/starts.

### macros: [key/value list]

In various configuration stanzas you may be allowed to use macros to replace text with dynamic values. Macros are **always** uppercase, and prefixed with two (2) percent-signs (%%) when used for expansion. See [Writing Handlers & Finishers](WritingHandlersAndFinishers.md) for information on how to use these in your own code.

Macros may be added or redefined at start time, via configuration. Macro values may contain other macros, but **only one level of macros will be expanded** (poor man's infinite-recursion avoidance).

As a rule, don't change a macro that defines a *value* to a macro containing another macro.

```yaml
macros:
  NAME: myserver
  CONFIGVERSION: 1.0
  SHORTVERSION: %%NAME/%%VERSION
  # The above works because all macros contained are defined
  # The below will not expand properly as it contains a macro that contains other macros
  FULLVERSION_BUSTED: %%SHORTVERSION %%CONFIGVERSION
  # The below works because all macros contained are defined
  FULLVERSION_WORKS: %%NAME/%%VERSION %%CONFIGVERSION
```

#### Defaults

- NAME: "JAR"
- VERSION: [JAR Version]
- SHORTVERSION: NAME/VERSION
- FULLVERSION: NAME/VERSION GoVersion MaxGoProcesses:NumberOfCPUs

### maxconnections: [number]

**Default: 0**
If 0, incoming connections to the listener are limited by the operating system, when it runs out of handles or ports. If > 0, provides a hard limit to the number of simultaneously accepted connections.

### requestidheadername: [header]

**Default: X-Requestid**
***Mandatory***
Sets the name of the request header to set with the request ID.

### slowrequestmax: [duration]

**Default: unset**
If set, requests taking longer than this will be logged to **slowlog**.

### tempfolder: [path]

**Default: /tmp**
Specifies the folder path where any temporary files should be stored. Temporary files will be put into sub-directories, and removed immediately after their useful life has ended, unless catastrophic failure occurs.

### timeout: [duration]

**Default: 0**
If 0, there is no timeout. *Ever*. If > 0, sets the global request/response timeout. A request lasting longer than this will be returned a **503 Service Unavailable**. It may be overridden on a per-Path basis via **Path.Timeout**.

### trustrequestidheader: [true|false]

**Default: false**
If set, will trust incoming request header **requestidheadername** to provide a proper requestId, otherwise will generate one and [over]write it.

### versionrequired: [version string]

**Default: [empty]**
Sets the minimum ``jar`` VERSION that the config is valid for. Fails on bootstrap if not met.

## Administration

### hotconfig: [true|false]

**Default: false**
If set, will trigger a graceful restart if the specified **config** file is modified.

### hotupdate: [true|false]

**Default: false**
If set, in conjunction with the **Update** Finisher, will allow a graceful restart onto new code after a successful update.

### updatepath: [url]

**Default: [empty]**
*WARNING:* Requires **ec2:true** or **keys** for AWS set in config/environment.
In conjunction with the **Update** Finisher, this specifies an *S3 URL* where a .zip containing an updated binary can be found (assumes running on an EC2 instance or in a credentialed environment with sufficient permission).

## TLS

JAR provides full support for TLS-fronted connections. Including transparently handling multiple certs/domains using SNI. As security standards have evolved, the minimum acceptable version is TLS v1.0 (RIP SSL).

### tls.certs [list of certs]

```yaml
 certs:
   "*.domain.com":
     keyfile: keys/domain.com/key.pem
     certfile: keys/domain.com/cert.pem
   "*.another.net":
     keyfile: keys/another.net/key.pem
     certfile: keys/anoter.net/cert.pem
```

### tls.ciphers [list]

**Default: [see example below for default]**
List of ciphersuites- ordered- that you want to support. The list of acceptable ciphersuites can be found in the Go [tls package documentation](https://golang.org/pkg/crypto/tls/#pkg-constants).

```yaml
 ciphers:
   - TLS_CHACHA20_POLY1305_SHA256
   - TLS_AES_256_GCM_SHA384
   - TLS_AES_128_GCM_SHA256
   - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
   - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
   - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
   - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
   - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
   - TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
   - TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
   - TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
   - TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
```

Available ciphers vary, but the superset is below (the provided hex ID is a reference to the RFC-assigned cipher suite ID)

```text
"TLS_RSA_WITH_RC4_128_SHA":                0x0005,
"TLS_RSA_WITH_3DES_EDE_CBC_SHA":           0x000a,
"TLS_RSA_WITH_AES_128_CBC_SHA":            0x002f,
"TLS_RSA_WITH_AES_256_CBC_SHA":            0x0035,
"TLS_RSA_WITH_AES_128_CBC_SHA256":         0x003c,
"TLS_RSA_WITH_AES_128_GCM_SHA256":         0x009c,
"TLS_RSA_WITH_AES_256_GCM_SHA384":         0x009d,
"TLS_ECDHE_ECDSA_WITH_RC4_128_SHA":        0xc007,
"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA":    0xc009,
"TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA":    0xc00a,
"TLS_ECDHE_RSA_WITH_RC4_128_SHA":          0xc011,
"TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA":     0xc012,
"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA":      0xc013,
"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA":      0xc014,
"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256": 0xc023,
"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256":   0xc027,
"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256":   0xc02f,
"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": 0xc02b,
"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384":   0xc030,
"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384": 0xc02c,
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305":    0xcca8,
"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305":  0xcca9,
"TLS_AES_128_GCM_SHA256":                  0x1301,
"TLS_AES_256_GCM_SHA384":                  0x1302,
"TLS_CHACHA20_POLY1305_SHA256":            0x1303,
```

### tls.enabled [true|false]

**Default: false**
If set, will process all of the TLS-related directives, building out a TLS config, and eventually using it. This will ignore **listen** in lieu of **tls.listen** (see **tls.httpredirects** below).

**NOTE:** Changing this value while running will create problems for graceful restarts, and should only be done during controlled stop/starts.

### tls.http2 [true|false] **EXPERIMENTAL**

**Default: false**
**EXPERIMENTAL**: If set, enables HTTP/2 protocol support, maybe. **EXPERIMENTAL**

### tls.httpredirects [true|false]

**Default: false**
Binds a simple HTTP listener to the **listen** address, and redirects all traffic there to the **tls.listen** address.

### tls.keepalivedisabled [true|false]

**Default: false**
If set, will disable keepalives when using TLS.

### tls.listen  [address]

**Default: :8443**
Specifies the *address:port* or *:port* to listen for HTTPS requests on.

**NOTE:** Changing this value while running will create problems for graceful restarts, and should only be done during controlled stop/starts.

### tls.maxversion [decimal version]

**Default: 1.3**
Acceptable values are (currently):

- 1.0
- 1.1
- 1.2
- 1.3

### tls.minversion [decimal version]

**Default: 1.2**
Acceptable values are (currently):

- 1.0
- 1.1
- 1.2
- 1.3

## Logging

### accesslog: [logfile]

**Default:  [standard out]**
Supplied **logfile** may be an absolute or relative file location to log structured "access log" information. The file need not exist, but the path must, and be writable by the executing user.

### commonlog: [logfile]

**Default: off**
Supplied **logfile** may be an absolute or relative file location to log structured "Apache common log" (combined-log style) information. The file need not exist, but the path must, and be writable by the executing user.

### debuglog: [logfile]

**Default:  [standard error]**
Only enabled if **debug: true** is set.
Supplied **logfile** may be an absolute or relative file location to log unstructured "debug log" information. The file need not exist, but the path must, and be writable by the executing user.

### errorlog: [logfile]

**Default: [standard error]**
Supplied **logfile** may be an absolute or relative file location to log unstructured "error log" information. The file need not exist, but the path must, and be writable by the executing user.

### logage: [days]

**Default: 28**
For any log logging to a file, specifies the rough maximum number of days to keep a rolled log.

### logbackups: [number]

**Default: 3**
For any log logging to a file, specifies the maximum number of rolled files to keep.

### logsize: [MBsize]

**Default: 100**
For any log logging to a file, specifies the rough maximum size (in megabytes) a log is allowed to get before rolling.

### slowlog: [logfile]

**Default: off**
Supplied **logfile** may be an absolute or relative file location to log requests that exceed **slowrequestmax**. The file need not exist, but the path must, and be writable by the executing user.

## Paths

A **Path** is a special thing to JAR. It is a simple structure that is akin to an Apache "Location", "VirtualHost", or "Dir" - all in one. Paths are *ordered* in the configuration, and when a request is being examined, the first matching Path is used to service the request.

```yaml
 paths:
   -
    Path: /nope
    Absolute: true
    Finisher: Forbidden
  -
    Path: /search
    Host: google.example.com
    Methods:
      - GET
      - POST
    Redirect: https://www.google.com%1
    Timeout: 60s
  -
    Path: /secure
    RateLimit: 100
    Allow: 192.168.0.0/16
    BasicAuthSource: files/secure.passwd
    BasicAuthRealm: Secure
    Pool: securePool
  -
    Path: /
    RateLimit: 100
    ForbiddenPaths:
      - ^\/secret
      - ^\/\d\d\d
    Handlers:
      - Recoverer
    Pool: default
```

### absolute: [true|false]

**Default: false**
If set, **path** is treated absolutely, instead of as a prefix.

### allow: [rules]

**Default: all**
Comma-delimited list of IP addresses/ranges to allow. Setting **allow** but not setting **deny** automatically flips **deny** to *all*. Any addresses in **allow** are evaluated before addresses in **deny**, except *all*.

```yaml
  -
    Path: /admin/restart
    Absolute: true
    Allow: 127.0.0.1
    # implicit Deny: all
    Finisher: Restart
```

### basicauthrealm: [string]

Name of the HTTP Auth Realm on this Path. Need not be unique. Should not be empty if basic auth is used

### basicauthsource: [url]

URL to specify where HTTP Basic Auth information should come from (file://). Setting this forces HTTP Basic Authentication for the Path.

### basicauthusers: [list]

**Default: all**
List of usernames allowed on this Path.

### bodybytelimit: [size in bytes]

If set, enables the BodyByteLimit handler on the path, enforcing any body is at most the specified size, or a *413 Request Entity too large* is returned.

### deny: [rules]

**Default: none**
Comma-delimited list of IP addresses/ranges to deny. Any addresses in **Allow** are evaluated before addresses in **Deny**.

```yaml
  -
    Path: /admin/update
    Deny: 137.143.0.0/16, 8.8.8.8
    Absolute: true
    Finisher: Update
```

### errorcode: [http code]

**Default: http.StatusOk (200)**
The HTTP response code that will be returned with ErrorMessage, IFF ErrorMessage is set.

### errormessage: [string]

**Default: none**
A static message to respond with, if this path is executed.

### finisher: [finisher name]

The name of a Finisher used to complete requests to this Path. Mutually exclusive to **pool** and **redirect**.

### forbiddenpaths: [list]

List of regular expressions that if matched against a URI path, will result in a 403. Barewords will match anywhere in the path. Remember to escape your fore-slashes. *Case-insensitive*. There is also the global **forbiddenpaths** config if that is more appropriate.

### handlers: [list]

This is a list of Handlers that will be applied to the Path, after any global handlers.

```yaml
Handlers:
  - Recoverer
  - Crashy
```

### headers: [list]

This is a list of request headers, in "name value" format, that will be used when matching the **Path**. They can be absolute, or use simple matching. Header names are assumed to be everything *left* of the first space. Values are assumed to be everything *right* of the first space. There is also the global **headers** config if that is more appropriate.

### hmacsigned: [true|false]

**Default: false**
If set, the request **path** must have a valid HMAC signature appended to it. the **Path** *must* also have at least **hmac.key** set in the **Options** (example below), which must be the key used to create the aforementioned HMAC signture. See the **HMACSigner** **Finisher** documentation below for more information.

```yaml
hmacsigned: true
Options:
   hmac.key: abcdefghijk123
   hmac.expiration: 3h
```

#### hmac.key: [string]

**Default: empty**
**REQUIRED**
Shared key to use between signer and verifier paths.

#### hmac.salt: [string]

**Default: empty**
Optional static salt to use, shared between signer and verifier paths.

#### hmac.expiration: [duration]

**Default: empty (off)**
If non-empty, must be a valid duration string. For signing, will automatically append a query parameter (before signature is computed) containing the offset expiration stamp. For verification, will be computed against the current system time if the signature is otherwise valid.

#### hmac.expirationfield: [string]

**Default: "expiration"**
Allows control over the name of the query parameter name used to store/retrieve the expiration stamp.

### host: [hostname|pattern]

**Default: [any]**
This is a hostname or hostname pattern that will be used when matching the Path. It can be absolute, or use expressions.

```yaml
Host: www.domain.com
```

```yaml
 # Match all hosts starting with 'dev-'
 Host: "{_:|^dev-.*}.domain.com"
```

### hosts: [list of hostname|pattern]

This is a list of hostnames or hostname patterns that the Path is valid for. See **Path.Host** for examples.

### methods: [list]

**Default: [any]**
This is a list of HTTP methods that will be used when matching the Path.

```yaml
Methods:
  - GET
  - HEAD
```

### name: [path name]

The name of a Path is used only in debug output, and is completely optional. If it is not set, the debug output will contain the ordered index number of the Path instead.

### options: [PathOptions]

A free-form field in a Path, used by specific Handlers or Finishers to consume path-specific configuration. This is added to the request Context, so keep it light.

```yaml
Options:
   MirrorRequest.Mirrors:
     - Pool1
     - Pool2
```

### path: [path]

Path is a URI path, starting with a forward-slash (/) and possibly with more specificity thereafter. By default, the path is treated as a prefix, thus */he* would match */he*, */help*, */helloooooo*, etc. If this is undesirable, set **absolute**. Without other configuration, this path will match any hostname, any method, any request that contains the path.

### pool: [pool name]

The name of a Pool used to complete requests to this Path. Mutually exclusive to **Finisher** and **Redirect**.

### ratelimit: [decimal requests/second]

RateLimit sets the number of requests-per-second-per-source (IP:port) allowed.

```yaml
RateLimit: 100.175
```

### ratelimitcollectonly: [boolean]

**Default: false**
If set, ratelimit violations will only print to the debug log, and not respond differently.

### ratelimitpurge: [duration]

**Default: 1 hour**
This is how long a ratelimit bucket can exist before it is expired.

### redirect: [url]

A URL to redirect requests to this Path. Mutually exclusive to **finisher** and **pool**. A macro *%1* may be put on the URL to substitute the request path.

```yaml
Redirect: https://www.google.com%1
```

### redirectcode: [http code]

**Default: 301**
By default, **redirect** requests use *301 Permanent*, but that can be changed here, within reason.

### redirecthostmatch: [regular expression]

Used with **redirect** to match groups, represented as `$1`, `$2`, etc. within the **redirect** value.

```yaml
  -
    Path: /
    Hosts:
      - "{_:|.*}.example.com"
    Redirect: https://$1.somewherelse.com%1
    RedirectHostMatch: "(.*).example.com"
```

### replacepath: [string]

Simple string replacement of this string, for the path string, before the request is proxied. No regexps.

```yaml
  -
    Path: /hello
    ReplacePath: /world
    Pool: default
```

### stripprefix: [prefix]

Removes a the specified string from the beginning of a URI path, before it is forwarded on. Useful when remapping e.g. */files/folder/thefile.html* to */folder/thefile.html*

```yaml
  -
    Path: /files
    StripPrefix: /files
    Pool: s3files
```

### timeout: [duration]

**Default: 0**
If > 0, sets the request/response timeout for this Path, overriding the global **timeout**. A request lasting longer than this will be returned a **503 Service Unavailable**.

## Pools

Pools are containers for one or more service endpoints providing analogous services, that are proxied, load-balanced, etc. Pool members are proxied differently depending on their protocol scheme. Currently ``https://``, ``http://``, ``s3://``, and ``ws://`` are supported. Not all configuration options are supported by all pool types.

```yaml
pools.healthcheckinterval: 30s
stickycookie.aes.ttl: 5m
pools:
  api:
    Name: api
    Members:
      - http://192.168.0.10:8080
      - http://192.168.0.11:8080
  workers:
    Name: wsworkers
    Sticky: true
    StickyCookieName: ROUTEID
    StickyCookieType: aes
    Members:
      - ws://192.168.0.10:8081
      - ws://192.168.0.11:8081
  echo:
    Name: echo
    HealthCheckURI: /
    Prune: true
    ConsistentHashing: true
    ConsistentHashSource: request
    ConsistentHashName: Host
    Members:
      - http://localhost:8081/
      - http://localhost:8082/
      - http://localhost:8083/

  default:
    Name: www
    Sticky: true
    Members:
      - http://192.168.0.9
```

### pools.defaultconsistenthashload: [float]

**Default: 1.25**
From the library: Load is used to calculate average load. See the code, the paper and Google's blog post to learn about it.

### pools.defaultconsistenthashpartitions: [integer]

**Default: 7**
From the library: Keys are distributed among partitions. Prime numbers are good to distribute keys uniformly. Select a big PartitionCount if you have too many keys.

### pools.defaultconsistenthashreplicationfactor: [integer]

**Default: 20**
From the library: Members are replicated on consistent hash ring. This number controls the number each member is replicated on the ring.

### pools.defaultmembererrorstatus: [healthcheckstatus]

**Default: Warning**
The default HealthCheckStatus for members in an error state. One of Unknown, Ok, Warning, or Critical. Overridden per-Pool setting **HealthCheckErrorStatus**.

### pools.defaultmemberweight: [number]

**Default: 1**
The default weight for a Pool member.

### pools.healthcheckinterval: [interval]

**Default: 1 minute**
Global for all pools. If set to 0, disables automatic healthchecks, otherwise sets the frequency of the healthchecks. This is most accurately described as a "maximum interval", by default, unless **healthcheckshotgun** its set.
**NOTE:** Healthchecking many members doesn't work so well on small, single-CPU boxes. You've been warned.

### pools.healthcheckshotgun: [true/false]

**Default: false**
Global for all pools. If set to false (the default), healthchecks will be adaptively scheduled over the interval, so for an interval of 1 minute, if there are 60 healthchecks to run, one will be fired off every second throughout that minute. The order of scheduling is not maintained from interval-to-interval.
If set to true, all healthchecks will be run concurrently at the beginning of each interval.

### pools.localmemberweight: [number]

**Default: 1000**
The weight for a Pool member who is AZ-local to the JAR instance.

### pools.prematerialize: [true/false]

**Default: false**
If set, each Pool will be materialized during bootstrap, instead of as-requested. Pools generally materialize very quickly, but a materialized Pool takes up more
memory (and goros) than a husk. Unless all of your Pools are used all of the time, leaving this alone is just fine.

### stickycookie.aes.ttl: [duration] (*experimental*)

**Defalt: 0 (off)**
Global for all pools. If set, and a pool has **Sticky** set, and **StickyCookieType: aes**, will embed an expiration in the encrypted cookie and verify it upon receipt.
This *WILL NOT* set the cookie-level expiration. This *should* be used with **Buffered** and **BufferedFails** set to at least *2*.
**CAUTION:**  This option is tagged experimental and should be used with caution.

### stickycookie.httponly: [true/false]

**Default: false**
Global for all pools. If set, and a pool has **Sticky** set, the cookie will have the [HTTPOnly](https://www.owasp.org/index.php/HttpOnly) property set.

### stickycookie.secure: [true/false]

**Default: false**
Global for all pools. If set, and a pool has **Sticky** set, the cookie will have the [Secure](https://www.owasp.org/index.php/SecureFlag) property set.

### buffered: [true|false]

**Default: false**
If set, all outgoing requests will be buffered. In the event of a timeout or 500-class error, the request will be retried on other members up to **bufferedfails**-minus-1 times.
**NOTE:** Buffered requests *may* be retried on members that failed previously. Do not use this, if that is unacceptable.
**ANOTHER NOTE:** This is generally a bad idea unless you are confident in the idempotence of your requests (or rather the endpoint handling your requests).

### bufferedfails: [number]

If **buffered** is set, this is the number of times a request may fail before giving up.

### consistenthashing: [true|false]

**Default: false**
If set, consistent hashing will be used on the pool, ensuring consistency and uniform distribution across pool members.

### consistenthashnames: [list of strings]

If **consistenthashing** is set, this value will be a list of fields whose values will be used as a hash key. **Must** be balanced with **consistenthashsources**!

### consistenthashsources: [list of header|cookie|request]

If **consistenthashing** is set, this value will be a list of sources to pull the value, specified by **consistenthashnames**, for the hash key.
For `header` and `cookie`, it is paired with **consistenthashnames** to choose which key from those maps is used.
For `request` it is paired with **consistenthashnames** to choose from one of `remoteaddr`, `host`, or `url`. For `remoteaddr` the source port is removed to keep the address stable. **Must** be balanced with **consistenthashnames**!

```yaml
pools:
  echo:
    prune: true
    HealthCheckURI: /
    ConsistentHashing: true
    ConsistentHashSources: request,request
    ConsistentHashNames: Host,remoteaddr
    Members:
      - http://localhost:8081/
      - http://localhost:8082/
      - http://localhost:8083/
```

### ec2affinity: [true|false]

**Default: false**
If set, and globally **ec2: true** then Pool member who are EC2 instances and in the same Availability Zone as the running JAR instance, will receive much higher
weight than other members.

### healthcheckdisabled: [true|false]

**Default: false**
If set, will disable healthchecks for members of this pool

### healthcheckerrorstatus: [healthcheckstatus]

**Default: pools.defaultmembererrorstatus**
The HealthCheckStatus for members in an error state. One of Unknown, Ok, Warning, or Critical.


config.go
package jar

import (
	"github.com/spf13/viper"

	"fmt"
	"os"
	"strings"
)

// Constants for configuration key strings
const (
	ConfigAccessLog            = ConfigKey("accesslog")
	ConfigAuthPool             = ConfigKey("authpool")
	ConfigCheckConfig          = ConfigKey("checkconfig")
	ConfigCommonLog            = ConfigKey("commonlog")
	ConfigDebug                = ConfigKey("debug")
	ConfigDebugLog             = ConfigKey("debuglog")
	ConfigDebugRequests        = ConfigKey("debugrequests")
	ConfigDebugResponses       = ConfigKey("debugresponses")
	ConfigDebugTimings         = ConfigKey("debugtimings")
	ConfigDumpConfig           = ConfigKey("dumpconfig")
	ConfigErrorLog             = ConfigKey("errorlog")
	ConfigHandlers             = ConfigKey("handlers")
	ConfigHotConfig            = ConfigKey("hotconfig")
	ConfigKeepaliveTimeout     = ConfigKey("keepalivetimeout")
	ConfigKeys                 = ConfigKey("keys")
	ConfigListen               = ConfigKey("listen")
	ConfigLogAge               = ConfigKey("logage")
	ConfigLogBackups           = ConfigKey("logbackups")
	ConfigLogSize              = ConfigKey("logsize")
	ConfigMaxConnections       = ConfigKey("maxconnections")
	ConfigPaths                = ConfigKey("paths")
	ConfigPools                = ConfigKey("pools")
	ConfigRequestIDHeaderName  = ConfigKey("requestidheadername")
	ConfigSlowLog              = ConfigKey("slowlog")
	ConfigSlowRequestMax       = ConfigKey("slowrequestmax")
	ConfigStripRequestHeaders  = ConfigKey("striprequestheaders")
	ConfigTempFolder           = ConfigKey("tempfolder")
	ConfigTimeout              = ConfigKey("timeout")
	ConfigTrustRequestIDHeader = ConfigKey("trustrequestidheader")
	ConfigAuthoritativeDomains = ConfigKey("authoritativedomains")
	ConfigVersionRequired      = ConfigKey("versionrequired")
	ConfigLogFakeXFF           = ConfigKey("fakexfflog")

	ConfigURLRouteHeaders            = ConfigKey("urlroute.enableheaders")
	ConfigURLRouteIDHeaderName       = ConfigKey("urlroute.idheadername")
	ConfigURLRouteEndpointHeaderName = ConfigKey("urlroute.endpointheadername")
	ConfigURLRouteNameHeaderName     = ConfigKey("urlroute.nameheadername")
	ConfigPoolHeaderName             = ConfigKey("urlroute.poolheadername")
	ConfigPoolMemberHeaderName       = ConfigKey("urlroute.poolmemberheadername")
)

var (
	// ConfigValidations is used to wire in func()error to be run, validating distributed configs
	ConfigValidations = make(map[string]func() error)
	// ConfigAdditions is used to wire in additional default configurations
	ConfigAdditions = make(configDefaultSetter)
)

// ConfigKey is a string type for static config key name consistency
type ConfigKey = string

// InitConfig creates an config, initialized with defaults and environment-set values, and returns it
func InitConfig() *viper.Viper {
	v := viper.New()

	v.AutomaticEnv()
	v.SetEnvPrefix("JAR")

	loadDefaults(v)

	return v
}

type configDefaultSetter map[string]interface{}

func (c *configDefaultSetter) Set(v *viper.Viper) {
	for k, val := range *c {
		v.SetDefault(k, val)
	}
}

// LoadConfig read the config file and returns a config object or an error
func LoadConfig(configFilename string, v *viper.Viper) error {

	if configFilename != "" {
		configFilenames := strings.Split(configFilename, ",")
		v.SetConfigFile(configFilenames[0])

		err := v.ReadInConfig()
		if err != nil {
			if _, ok := err.(viper.ConfigParseError); ok {
				return err
			}
			return fmt.Errorf("unable to locate config file '%s': %w", configFilenames[0], err)
		}
		for _, configFile := range configFilenames[1:] {
			file, err := os.Open(configFile) // For read access.
			if err != nil {
				return fmt.Errorf("unable to open config file '%s': %w", configFile, err)
			}
			defer file.Close()
			if err = v.MergeConfig(file); err != nil {
				return fmt.Errorf("unable to parse/merge Config file '%s': %w", configFile, err)
			}
		}
	}

	return nil
}

// loadDefaults sets all the default values for the config. An error may be returned, to support future operations.
func loadDefaults(v *viper.Viper) error {
	v.SetDefault(ConfigDebug, false)                            // Enable vociferous output
	v.SetDefault(ConfigDebugRequests, false)                    // Enable vociferous output of requests
	v.SetDefault(ConfigDebugResponses, false)                   // Enable vociferous output of responses
	v.SetDefault(ConfigListen, ":8080")                         // ip:port or :port to listen on
	v.SetDefault(ConfigAccessLog, "")                           // Path to file where accesslog, else stdout
	v.SetDefault(ConfigDebugLog, "")                            // Path to file where debug should log to, else stderr
	v.SetDefault(ConfigErrorLog, "")                            // Path to file where errorlog should log to, else stderr
	v.SetDefault(ConfigSlowLog, "")                             // Path to file where the slow requests should log to
	v.SetDefault(ConfigLogSize, 100)                            // Maximum size, in MB, that the currently log can be before rolling
	v.SetDefault(ConfigLogBackups, 3)                           // Maximum number of rolled logs to keep
	v.SetDefault(ConfigLogAge, 28)                              // Maximum age, in days, to keep rolled logs
	v.SetDefault(ConfigPaths, make([]interface{}, 0))           // paths are blocks to define URI prefixes, and the handlers they should use
	v.SetDefault(ConfigPools, make([]interface{}, 0))           // pools are blocks to define loadbalancer pools, their configuration, and their members
	v.SetDefault(ConfigKeys, make(map[string]string))           // keys[keyname] = base64-encodedkey
	v.SetDefault(ConfigAuthPool, make(map[string]interface{}))  // authpool[poolname] = map[opt]value
	v.SetDefault(ConfigHandlers, make([]string, 0))             // handlers is a list of global handlers
	v.SetDefault(ConfigHotConfig, false)                        // hotconfig controls whether we watch for config changes, and reload on change, or not
	v.SetDefault(ConfigTempFolder, "/tmp")                      // tempfolder is where temporary files go
	v.SetDefault(ConfigStripRequestHeaders, make([]string, 0))  // striprequestheaders lists request headers that should be removed
	v.SetDefault(ConfigTimeout, 0)                              // Sets the total request/response timeout. Can be overridden per-path
	v.SetDefault(ConfigKeepaliveTimeout, "5s")                  // Sets the amount of time keptalive sockets linger
	v.SetDefault(ConfigMaxConnections, 0)                       // Maximum number of simultaneous connections to the listener
	v.SetDefault(ConfigRequestIDHeaderName, "X-Request-ID")     // Name of the header to use for adding the requestid
	v.SetDefault(ConfigSlowRequestMax, "")                      // Sets the default slow request time for logging
	v.SetDefault(ConfigTrustRequestIDHeader, false)             // Enable trusting of incoming ConfigRequestIDHeaderName to set the requestid
	v.SetDefault(ConfigAuthoritativeDomains, make([]string, 0)) // If set, restricts request handling to the domains listed, otherwise 400s
	v.SetDefault(ConfigVersionRequired, "")                     // Sets the minimum ``go-jar`` VERSION that the config is valid for.

	v.SetDefault(ConfigURLRouteHeaders, false)                       // Enables setting the headers below, if SwitchHandler is used
	v.SetDefault(ConfigURLRouteIDHeaderName, "X-URLROUTEID")         // Name of the header to use for route ID
	v.SetDefault(ConfigURLRouteEndpointHeaderName, "X-URLROUTEDEST") // Name of the header to use for route endpoint
	v.SetDefault(ConfigURLRouteNameHeaderName, "X-URLROUTENAME")     // Name of the header to use for route name
	v.SetDefault(ConfigPoolHeaderName, "X-URLPOOL")                  // Name of the header to use when capturing the servicing pool
	v.SetDefault(ConfigPoolMemberHeaderName, "X-URLPOOLMEMBER")      // Name of the header to use when capturing the servicing request pool member
	ConfigAdditions.Set(v)

	return nil
}

// ValidateExtras runs through a list of referenced functions, and returns any errors they return.
// All functions will be run, so an array of errors may be returned
func ValidateExtras() []error {
	errs := make([]error, 0)

	for _, f := range ConfigValidations {
		err := f()
		if err != nil {
			errs = append(errs, err)
		}
	}

	return errs
}
